doc/kubernetes/modules/ROOT/pages/running/deployments/active-passive-sync.adoc

```diff
@@ -0,0 +1,225 @@
+
+[%autowidth]
+|===
+| Failure | Recovery | RPO^1^ | RTO^2^
+
+| Database node
+| If the writer instance fails, the database can promote a reader instance in the same or other datacenter to be the new writer.
+| No data loss
+| Seconds to minutes (depending on the database)
+
+| Keycloak node
+| Multiple Keycloak instances run in each datacenter. If one instance fails, it takes a few seconds for the other nodes to notice the change, and some incoming requests might receive an error message or are delayed for some seconds.
+| No data loss
+| Less than one minute
+
+| Infinispan node
+| Multiple Infinispan instances run in each datacenter. If one instance fails, it takes a few seconds to the other nodes to notice the change. Sessions are stored in at least two Infinispan nodes, so a single node failure doesn't lead to data loss.
+| No data loss
+| Less than one minute
+
+| Infinispan cluster failure
+| If the Infinispan cluster fails in the active datacenter, Keycloak won't be able to send session data to the secondary datacenter.
+The state between Keycloak and Infinispan is out-of-sync, and also the state between the two Datacenters.
+Keycloak will continue on the best effort basis, but the service might be degraded due to retry mechanisms.
+Even when the Infinispan cluster is restored, its data will be out-of-sync with Keycloak.
+
+Manual switchover to the secondary datacenter is recommended. As that datacenter is out-of-sync, the customer should consider if all data should be cleared from the session store of the passive datacenter to avoid out-of-date information.
+| Loss of service and data
+| Human intervention required
+
+| Connectivity Infinispan
+| If the connectivity between the two datacenters is lost, session information can't be sent to the other datacenter.
+Incoming requests might receive an error message or are delayed for some seconds.
+The primary site marks the secondary site offline, and will stop sending data to the secondary.
+The setup is degraded until the connection is restored and the session data is re-synchronized to the secondary datacenter.
+| No data loss ^3^
+| Less than one minute
+
+| Connectivity Database
+| If the connectivity between the two datacenters is lost, the synchronous replication will fail, and it might take some time for the primary site to mark the secondary offline.
+Some requests might receive an error message or are delayed for some seconds.
+Manual operations might be necessary depending on the database.
+| No data loss ^3^
+| Seconds to minutes (depending on the database)
+
+| Primary Datacenter
+| If none of the Keycloak nodes is available, the loadbalancer will detect the outage and redirect the traffic to the secondary site.
+Some requests might receive an error message while the loadbalancer hasn't detected the primary datacenter failure.
+The setup will be degraded until the primary site is back up and the session state has been manually synced from the secondary to the primary site.
```

Comment:
Can it be manual? I don't think there is any scenario where we want to reinstate the primary automatically. At least, Infinispan needs to be synchronized manually, and probably the DB needs to manually switch back to the primary.

Reply:
> probably the DB needs to manually switch back to the primary.

Aurora DB will continue to function without any manual intervention, however the writer will be on a different AZ to the primary which will add some latency.

> Can it be manual?

Not out of the box. We would need to add some kind of Lambda to update the Route53 records.

I think there are 3 possible courses of action here:

1. Update health probes to take into account xsite state so that failback can't occur
  - https://github.com/keycloak/keycloak-benchmark/issues/534
  
2. Implement a Lambda to update Route53 records in the event of a failover so that failback is not possible

3. Utilise Global Accelerator, with a Lambda to flip the weights of the Primary and Backup endpoint on failure to prevent the primary serving traffic. This would also overcome the issue that me and @ahus1 discovered with Accelerator failover yesterday.

For both 2/3 we will need to document in the operational guide how to failback to the Primary manually when the time is right.

IMO 1. is out of scope for the initial implementation as it will require changes to Keycloak itself. The work required for 2/3 is pretty similar, but given the downsides of the DNS based routing, my preference would be to implement 3.

Reply:
Option 4 will be to prevent the Keycloak pods from starting before Infinispan is in sync. :grinning:

Reply:
> Option 4 will be to prevent the Keycloak pods from starting before Infinispan is in sync. 😀

I think that logic would have to be added to the Keycloak Operator, so that it could scale down the StatefulSet to 0 instances until Infinispan was synced. If we just had this check on Keycloak pod start, or as part of the health check used by the Pod, then the pod will be restarted and eventually hit the pod restart limit if Infinispan wasn't synced in time.

I think for such an approach to work, we would potentially need a different health check for the pod and LoadBalancer as they have different requirements.

Reply:
I would be ok to have an instruction in the docs for now when we stay with Route 53: "Before you bring up the primary site again, disable it in manually in the loadbalancer first" - This is not ideal and we would hopefully improve it soon one way or another, still it is a start.

Reply:
This doesn't seem very resilient to me.  A Keycloak  cluster could go down due to K8s node failure(s), in which case the pod's will automatically restart on a different node. From a Route53 perspective this would mean that we could switch from Primary->Backup->Primary without user input.

Reply:
A single K8S node should not bring down Infinispan or Keycloak deployment, so I was more thinking about a catastrophic OpenShift failure here. In a next iteration we might add (weak) pod-anti-affinity configurations to the pods to minimize a node failure taking down more than one pod of Keycloak or Infinispan. 

Good luck with the automation. If it doesn't work as expected, just be aware that I would still accept a smaller solution in a first iteration.